### ISSUE TITLE
Unexpected values in prediction mask

### ISSUE THREAD

**Issue by 01bui** [ultralytics/ultralytics#5368]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello YOLOv8 Team,

I've been working with the yovov8m-seg model to segment a publicly available chest x-ray dataset. While most of the predictions are satisfactory, I've encountered some anomalies in the prediction masks. Given that this is a binary segmentation task, I anticipated mask values of either 0 or 255. However, that's not consistently the case.

Here are some representative results:

Expected mask outcomes from the trained model:
![image](https://github.com/ultralytics/ultralytics/assets/8173465/c4bb6555-86ab-490e-8b19-4d2fd9a1c84b)
![image](https://github.com/ultralytics/ultralytics/assets/8173465/00c1aa8e-e5bc-4867-bc66-5f1a40980b9f)

Unexpected mask outcomes:
![image](https://github.com/ultralytics/ultralytics/assets/8173465/81985e55-e321-48c3-b23e-c665c77ef839)
![image](https://github.com/ultralytics/ultralytics/assets/8173465/df7a2d30-ab8b-45b0-b984-742d29b720cb)

Furthermore, a histogram of the prediction mask revealed the presence of these unexpected values:
![image](https://github.com/ultralytics/ultralytics/assets/8173465/64b07744-30b8-4233-846d-c3fb04d419a1)

Would you please provide insights to troubleshoot this issues?

Thank you.

**Comment by zshariff6506** [comment]
This looks interesting. Can you give me a snippet of your code or explain how you took the yolov8m-seg to the white and black images?

**Comment by zshariff6506** [comment]
@zshariff6506

**Comment by glenn-jocher** [comment]
@01bui thank you for bringing this to our attention. It's interesting to see these unexpected values in your mask outcomes. 

The behaviour you're seeing is actually part of how the Ultralytics YOLOv8 segmentation models are designed to work. These models, such as yolov8m-seg, are not binary classifiers. Instead, they generate output through a Softmax function that triggers probabilities for each class for each pixel in the image. In a binary segmentation task, the model assigns two probabilities per pixel: one for class A (let's say 0) and one for class B (let's say 255). These probabilities are normalized in a way that their sum equals 1. 

If we look at your outputs, the expected binary values of 0 and 255 signify high confidence in belonging to either class A or B respectively. In moments of less certainty, output values can be somewhere in between, as reflected in your unexpected mask outcomes. Remember, the model quantifies its confidence in assigning a pixel to a specific class, which allows it to convey its uncertainty about certain areas of the image.

If your target is to achieve black & white segmentation maps, one viable approach would be to apply a threshold to these prediction probabilities. In binary cases, a typical choice is to assign pixels with probabilities greater than 0.5 to class B and those with lower probabilities to class A. 

Keep in mind, fine-tuning this threshold can help improve your results specifically for your application. And always be conscious that in doing so, you might be introducing more false positives or false negatives, depending on which way you adjust the threshold. 

I hope this clears up your confusion. If you have any further questions, feel free to ask!

**Comment by 01bui** [comment]
@glenn-jocher Thanks for your reply. How to I get the probabilities instead of the mask?

**Comment by glenn-jocher** [comment]
Hello @01bui,

Great question!

To get the probabilities directly, you can modify the post-processing step of your pipeline. 

Currently, your pipeline likely includes a step where the output of the Softmax function is converted into your final segmentation mask. This step may involve choosing the class with the highest probability for each pixel or applying a threshold, as I suggested before.

If you bypass this step, the model's direct output should be a 3D tensor where the third dimension represents the normalized class probabilities for each pixel. This tensor contains the probability distribution produced by the Softmax function across your classes for each pixel, and it is effectively what you're after.

Just remember, the model's output, before any post-processing, gives you these class probabilities. If you want to work with these probabilities directly, make sure to intercept the output before it gets binarized into a mask.

Please don't hesitate to get in touch if you need more clarity or further assistance.

**Comment by 01bui** [comment]
@glenn-jocher 
I'm curious about how to obtain probabilities from this code. According to the documentation, r.masks provides the label for each pixel.
```
results = model(img)
   for r in results:
      im_array = r.masks.data.cpu().numpy()
```
